lib/transports/ws.js

```diff
@@ -0,0 +1,269 @@
+const WebSocket = require('ws')
+const EventEmitter = require('events')
+const crypto = require('crypto')
+
+const WS_HOST = 'wss://max-ws.maicoin.com'
+const KEEP_ALIVE_INTERVAL = 3000
+
+/* Communicate with MAX Exchange Websocket API */
+class WSV1 extends EventEmitter {
+  /**
+   * Initialize a websocket transport class.
+   *
+   * @constructor
+   * @param {Object} options
+   * @param {string} [options.accessKey]
+   * @param {string} [options.secretKey]
+   */
+  constructor (options = {
+    accessKey: '', secretKey: ''
+  }) {
+    super()
+
+    this._accessKey = options.accessKey
+    this._secretKey = options.secretKey
+
+    this._isOpen = false
+    this._isAlive = false
+    this._authenticated = false
+
+    this._ws = null
+    this._intervalId = null
+  }
+
+  /**
+   * Opens a connection to the websocket API server. Will reject open if websocket is already
+   * connected.
+   *
+   * @return {Promise}
+   */
+  open () {
+    if (this._isOpen || this._ws !== null) {
+      return Promise.reject(new Error('Websocket is already open.'))
+    }
+
+    this._ws = new WebSocket(WS_HOST)
+
+    this._ws.on('open', this._handleOpen.bind(this))
+    this._ws.on('open', this._handleConnection.bind(this))
+    this._ws.on('message', this._handleMessage.bind(this))
+    this._ws.on('close', this._handleClose.bind(this))
+    this._ws.on('error', this._handleError.bind(this))
+
+    return new Promise((resolve) => {
+      this._ws.once('open', () => {
+        resolve()
```

Comment:
reject promise if websocket construction failed

Reply:
https://github.com/maicoin/max-exchange-api-node/pull/7/files/6b1b2b6268bc3d7221aeb78d5892d8ee97f531f4#diff-4f615e0c741901bd8d4721c5a4654b41R42 
already handled here.

Other cases should be handle by events.